packages/malloc/src/pool.ts

```diff
@@ -10,42 +10,107 @@ import {
 } from "./api";
 import { wrap } from "./wrap";
 
+const STATE_TOP = 0;
+const STATE_END = 1;
+const STATE_FREE = 2;
+const STATE_USED = 3;
+
+const STATE_NUM_FIELDS = 4;
+const SIZEOF_STATE = STATE_NUM_FIELDS * Uint32Array.BYTES_PER_ELEMENT;
+
+
 export class MemPool implements IMemPool {
     buf: ArrayBuffer;
-    protected top: number;
-    protected start: number;
-    protected end: number;
+
+    /**
+     * Start is the anchor index inside the arraybuffer, so we can't save it inside the arraybuffer itself.
+     * If you pass the ArrayBuffer to other consumers they must use the same start value
+    */
+    protected readonly start: number;
+
     protected doCompact: boolean;
     protected doSplit: boolean;
     protected minSplit: number;
-    protected _free: MemBlock | null;
-    protected _used: MemBlock | null;
-
     protected u8: Uint8Array;
+    protected u32: Uint32Array;
 
     constructor(opts: Partial<MemPoolOpts> = {}) {
         this.buf = opts.buf ? opts.buf : new ArrayBuffer(opts.size || 0x1000);
         this.u8 = new Uint8Array(this.buf);
-        this.start = opts.start != null ? align(Math.max(opts.start, 8), 8) : 8;
-        this.end =
-            opts.end != null
-                ? Math.min(opts.end, this.buf.byteLength)
-                : this.buf.byteLength;
-        if (this.start >= this.end) {
-            illegalArgs(
-                `invalid address range (0x${this.start.toString(
-                    16
-                )} - 0x${this.end.toString(16)})`
-            );
-        }
-        this.top = this.start;
         this.doCompact = opts.compact !== false;
         this.doSplit = opts.split !== false;
         this.minSplit = opts.minSplit || 16;
-        this._free = null;
-        this._used = null;
+
+        this.start = (opts.start != null ? align(Math.max(opts.start, 8), 8) : 8);
+        this.u32 = new Uint32Array(this.buf, this.start);
+
+        if (!opts.skipInitialization) {
+            const resolvedEnd = opts.end != null
+                ? Math.min(opts.end, this.buf.byteLength)
+                : this.buf.byteLength;
+
+            if (this.start >= resolvedEnd) {
+                illegalArgs(
+                    `invalid address range (0x${this.start.toString(
+                        16
+                    )} - 0x${resolvedEnd.toString(16)})`
+                );
+            }
+
+            this.end = resolvedEnd;
+            this.top = this.start + SIZEOF_STATE;
+            this._free = null;
+            this._used = null;
+        }
+    }
+
+
+
+    protected get end(): number {
+        return this.u32[STATE_END];
```

Comment:
I see what you did here. You're using `u32` also for what I had as `state`, the other Uint32Array which was just mapping the allocator state fields and hence you added the byte offset when you create it... IMHO keeping state separately is somewhat more clear, but let you decide...

Reply:
Oki :)